File: src/lib/severity.ts
// Cross-scanner severity normalization. Port of api/src/sec_dashboard/severity.py.
//
// Order: CRITICAL=5 > HIGH=4 > MEDIUM=3 > LOW=2 > INFO=1 > SUPPRESSED=0.

export enum Severity {
  Critical = 5,
  High = 4,
  Medium = 3,
  Low = 2,
  Info = 1,
  Suppressed = 0,
}

export type SeverityLabel = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'INFO' | 'SUPPRESSED';

export const SEVERITY_ORDER: Severity[] = [
  Severity.Critical,
  Severity.High,
  Severity.Medium,
  Severity.Low,
  Severity.Info,
  Severity.Suppressed,
];

export function severityLabel(s: Severity): SeverityLabel {
  switch (s) {
    case Severity.Critical:
      return 'CRITICAL';
    case Severity.High:
      return 'HIGH';
    case Severity.Medium:
      return 'MEDIUM';
    case Severity.Low:
      return 'LOW';
    case Severity.Info:
      return 'INFO';
    case Severity.Suppressed:
      return 'SUPPRESSED';
  }
}

export function severityColor(s: Severity): string {
  switch (s) {
    case Severity.Critical:
      return '#b71c1c';
    case Severity.High:
      return '#e64a19';
    case Severity.Medium:
      return '#f9a825';
    case Severity.Low:
      return '#1976d2';
    case Severity.Info:
      return '#616161';
    case Severity.Suppressed:
      return '#9e9e9e';
  }
}

export function fromTrivy(raw: string | undefined): Severity {
  const s = (raw || '').toUpperCase();
  switch (s) {
    case 'CRITICAL':
      return Severity.Critical;
    case 'HIGH':
      return Severity.High;
    case 'MEDIUM':
      return Severity.Medium;
    case 'LOW':
      return Severity.Low;
    case 'UNKNOWN':
      return Severity.Info;
    default:
      return Severity.Info;
  }
}

export function emptyCounts(): Record<SeverityLabel, number> {
  return { CRITICAL: 0, HIGH: 0, MEDIUM: 0, LOW: 0, INFO: 0, SUPPRESSED: 0 };
}

File: backend/src/sec_dashboard/severity.py
"""Cross-scanner severity normalization.

The user requires every result ranked CRITICAL > HIGH > MEDIUM > LOW > INFO,
even though each tool has its own scale. Mapping rules are documented in the
plan §3 / severity normalization table.
"""

from enum import IntEnum


class Severity(IntEnum):
    CRITICAL = 5
    HIGH = 4
    MEDIUM = 3
    LOW = 2
    INFO = 1
    # SUPPRESSED is a rebucket target for findings on the allowlist (see
    # scans/suppressions.py). The raw finding is preserved — only its
    # severity_normalized changes — so the dashboard can render an
    # "accepted risk" panel without dropping data.
    SUPPRESSED = 0

    @property
    def label(self) -> str:
        return self.name


SEVERITY_ORDER = [
    Severity.CRITICAL,
    Severity.HIGH,
    Severity.MEDIUM,
    Severity.LOW,
    Severity.INFO,
    Severity.SUPPRESSED,
]


def from_trivy(s: str) -> Severity:
    s = (s or "").upper()
    return {
        "CRITICAL": Severity.CRITICAL,
        "HIGH": Severity.HIGH,
        "MEDIUM": Severity.MEDIUM,
        "LOW": Severity.LOW,
        "UNKNOWN": Severity.INFO,
    }.get(s, Severity.INFO)


def from_kube_bench(status: str, profile: str) -> Severity:
    """kube-bench has no severity field; weight by profile + status."""
    status = (status or "").upper()
    profile = (profile or "").lower()
    if status == "FAIL":
        if profile in {"master", "controlplane", "etcd"}:
            return Severity.HIGH
        return Severity.MEDIUM
    if status == "WARN":
        return Severity.MEDIUM
    if status == "INFO":
        return Severity.INFO
    return Severity.INFO


def from_kubescape(severity_label: str, score: float | None = None) -> Severity:
    s = (severity_label or "").lower()
    if s == "critical":
        return Severity.CRITICAL
    if s == "high":
        return Severity.HIGH
    if s == "medium":
        return Severity.MEDIUM
    if s == "low":
        return Severity.LOW
    if score is not None:
        if score >= 9:
            return Severity.CRITICAL
        if score >= 7:
            return Severity.HIGH
        if score >= 4:
            return Severity.MEDIUM
        return Severity.LOW
    return Severity.INFO


def from_lynis(kind: str, impact: str | None = None) -> Severity:
    """Lynis warnings may carry IMPACT=high|medium|low; suggestions → LOW."""
    kind = (kind or "").lower()
    impact = (impact or "").lower()
    if kind == "warning":
        if impact == "high":
            return Severity.HIGH
        if impact == "low":
            return Severity.LOW
        return Severity.MEDIUM
    if kind == "suggestion":
        return Severity.LOW
    return Severity.INFO
